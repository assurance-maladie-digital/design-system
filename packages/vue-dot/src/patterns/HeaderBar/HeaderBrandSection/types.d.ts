export interface LogoInfo {
	src: string;
	alt: string;
}

export interface Service {
	mainTitle: string;
	highlight: string;
	subTitle: string;
}
