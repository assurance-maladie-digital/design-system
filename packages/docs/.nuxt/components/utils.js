// nuxt/nuxt.js#8607
export function wrapFunctional(options) {
  if (!options || !options.functional) {
    return options
  }

  const propKeys = Array.isArray(options.props) ? options.props : Object.keys(options.props || {})

  return {
    render(h) {
      const attrs = {}
      const props = {}

      for (const key in this.$attrs) {
        if (propKeys.includes(key)) {
          props[key] = this.$attrs[key]
        } else {
          attrs[key] = this.$attrs[key]
        }
      }

      return h(options, {
        on: this.$listeners,
        attrs,
        props,
        scopedSlots: this.$scopedSlots,
      }, this.$slots.default)
    }
  }
}
