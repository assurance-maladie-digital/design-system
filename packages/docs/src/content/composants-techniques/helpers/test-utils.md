---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
