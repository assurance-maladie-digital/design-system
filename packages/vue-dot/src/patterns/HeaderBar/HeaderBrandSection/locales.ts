export const locales = {
	logoCnam: 'Caisse nationale',
	logoAmeli: 'Ameli.fr',
	logoAmeliPro: 'AmeliPro'
};
