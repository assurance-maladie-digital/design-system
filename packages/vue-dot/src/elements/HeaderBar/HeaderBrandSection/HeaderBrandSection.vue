<template>
	<div class="d-flex">
		<img
			:src="headerLogo"
			alt="Caisse Nationale d'Assurance Maladie"
		>
		<div
			v-if="hasDaughterBrand"
			class="d-flex align-center"
		>
			<slot />
		</div>
		<div
			v-else-if="theme === 'ameli.fr' || theme === 'ameli-pro'"
			class="d-flex"
		>
			<img
				:src="dividerModel"
				alt="Caisse Nationale d'Assurance Maladie"
			>
			<img
				:src="daughterBrand"
				alt="Caisse Nationale d'Assurance Maladie"
			>
		</div>
		<div
			v-if="displayServiceSection"
			:style="`color: ${themeColor};`"
			class="d-flex align-center"
		>
			<img
				:src="dividerModel"
				alt="Caisse Nationale d'Assurance Maladie"
			>
			<div class="d-flex flex-column">
				<div class="vd-header-title-main">
					{{ serviceTitle }}
				</div>
				<div class="vd-header-title-sub">
					{{ serviceSubTitle }}
				</div>
			</div>
		</div>
	</div>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { customizable } from '../../../mixins/customizable';

	import { HeaderBarConfig } from '../types';

	const Props = Vue.extend({
		props: {
			hasDaughterBrand: {
				type: Boolean,
				default: false
			},
			headerConfig: {
				type: Object as PropType<HeaderBarConfig>,
				default: undefined
			},
			theme: {
				type: String,
				default: 'ameli.fr',
				validator(value: string): boolean {
					const isValid = value.match(/^(cnam|ameli.fr|ameli-pro|risque-pro|custom)$/) !== null;
					if (!isValid) {
						// eslint-disable-next-line no-console
						console.error(`Wrong value for the \`type\` prop. 
						Given: "${value}", expected "(cnam|ameli.fr|ameli-pro|risque-pro|custom)".`);
					}
					return true;
				}
			},
			isMobileScreen: {
				type: Boolean,
				default: false
			},
			themeColor: {
				type: String,
				default: '#0c419a'
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component({
		inheritAttrs: false
	})
	export default class HeaderBrandSection extends MixinsDeclaration {
		// translations
		locales = locales;

		get imageFolder(): string {
			return this.isMobileScreen ? 'mobile' : 'desktop';
		}

		get daughterBrand(): string {
			let img :string;

			if(this.theme === 'ameli-pro') {
				img = 'amelipro';
			} else if(this.theme === 'ameli.fr') {
				img = 'ameli';
			} else {
				img = 'no-image';
			}

			return require(`../../../assets/marque-fille/${this.imageFolder}/${img}.svg`);
		}

		get dividerModel(): string {
			let img :string;

			if(this.theme === 'ameli-pro') {
				img = 'amelipro';
			} else {
				img = 'bleu-am';
			}

			return require(`../../../assets/dividers/${this.imageFolder}/${img}.svg`);
		}

		get headerLogo(): string {
			let img :string;

			if(this.theme === 'risque-pro') {
				img = 'risquepro';
			} else {
				img = 'assurance-maladie';
			}

			return require(`../../../assets/logos/${this.imageFolder}/${img}.svg`);
		}

		get displayServiceSection(): boolean {
			if(this.headerConfig?.service) {
				if(!this.headerConfig.service.mobileDisplay) {
					return this.isMobileScreen ? false : true;
				}
				return true;
			}
			return false;
		}

		get displayServiceTitle(): boolean {
			if(this.headerConfig?.service?.title) {
				if(!this.headerConfig.service.title.mobileDisplay) {
					return this.isMobileScreen ? false : true;
				}
				return true;
			}
			return false;
		}

		get displayServiceSubTitle(): boolean {
			if(this.headerConfig?.service?.subTitle) {
				if(!this.headerConfig.service.subTitle.mobileDisplay) {
					return this.isMobileScreen ? false : true;
				}
				return true;
			}
			return false;
		}

		get serviceTitle(): string {
			return this.headerConfig?.service?.title?.text ? this.headerConfig.service.title.text as string : '';
		}

		get serviceSubTitle(): string {
			return this.headerConfig?.service?.subTitle?.text ? this.headerConfig.service.subTitle.text as string : '';
		}
	}
</script>

<style lang="scss" scoped>
	.vd-header {
		&-brand {
			padding: 27px;

			&-responsive {
				padding: 8px;
			}
		}

		&-title {
			&-main {
				font-size: 16px;
				line-height: 18px;
			}

			&-sub {
				font-size: 12px;
				line-height: 14px;
			}
		}
	}
</style>
