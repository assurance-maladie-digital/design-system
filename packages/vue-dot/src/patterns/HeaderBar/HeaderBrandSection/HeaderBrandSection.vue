<template>
	<VSheet
		:height="height"
		class="vd-header-brand-section d-flex"
	>
		<component
			:is="logoContainerComponent"
			:aria-current-value="null"
			:aria-label="locales.homeLinkLabel"
			:to="homeLink"
			class="vd-header-home-link"
		>
			<Logo
				:hide-signature="hideSignature"
				:risque-pro="isRisquePro"
				:size="logoSize"
			/>
		</component>

		<slot>
			<svg
				v-if="showDivider"
				:width="dividerDimensions.width"
				:height="dividerDimensions.height"
				:fill="dividerColor"
				role="img"
				focusable="false"
				aria-hidden="true"
				xmlns="http://www.w3.org/2000/svg"
				viewBox="0 0 22 64"
			>
				<path d="M14.3 49.3c-.2 0-.4-.2-.4-.4V14.2c0-.2.2-.4.4-.4.3 0 .5.2.5.4v34.7c0 .2-.2.4-.5.4Z" />
			</svg>

			<img
				v-if="secondaryLogo"
				:src="secondaryLogo.src"
				:alt="secondaryLogo.alt"
			>

			<div
				v-else-if="service.title || service.subTitle"
				class="d-flex justify-center flex-column primary--text"
			>
				<h1
					v-if="service.title"
					:class="{ 'vd-compte-entreprise-title': isCompteEntreprise }"
					class="vd-header-title text-caption text-md-subtitle-1 font-weight-medium"
				>
					<template v-if="isCompteEntreprise">
						{{ service.title.text }}
						<span>{{ service.title.highlight }}</span>
					</template>

					<template v-else>
						{{ service.title }}
					</template>
				</h1>

				<h2
					v-if="showServiceSubTitle"
					class="vd-header-title text-caption"
				>
					{{ service.subTitle }}
				</h2>
			</div>
		</slot>
	</VSheet>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { tokens } from '@cnamts/design-tokens';

	import { Dimensions, Next } from '../../../types';
	import { LogoSizeEnum } from '../../../elements/Logo/LogoSizeEnum';

	import { ThemeEnum } from '../ThemeEnum';

	import { LogoInfo, Service } from './types';
	import { locales } from './locales';
	import { secondaryLogoMapping } from './secondaryLogoMapping';
	import { dividerDimensionsMapping } from './dividerDimensionsMapping';

	const Props = Vue.extend({
		props: {
			theme: {
				type: String as PropType<ThemeEnum>,
				required: true
			},
			serviceTitle: {
				type: String,
				default: undefined
			},
			serviceSubTitle: {
				type: String,
				default: undefined
			},
			mobileVersion: {
				type: Boolean,
				default: false
			},
			homeLink: {
				type: [String, Boolean, Object] as PropType<Next>,
				default: '/'
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component
	export default class HeaderBrandSection extends MixinsDeclaration {
		locales = locales;

		get service(): Service {
			if (this.theme === ThemeEnum.COMPTE_ENTREPRISE) {
				const { title, subTitle } = locales.compteEntreprise;

				return {
					title,
					subTitle
				};
			}

			return {
				title: this.serviceTitle,
				subTitle: this.serviceSubTitle
			};
		}

		get height(): string {
			if (this.mobileVersion && this.hasSecondaryLogo) {
				return '32px';
			}

			return this.mobileVersion ? '40px' : '64px';
		}

		get isRisquePro(): boolean {
			return this.theme === ThemeEnum.RISQUE_PRO;
		}

		get isCompteEntreprise(): boolean {
			return this.theme === ThemeEnum.COMPTE_ENTREPRISE;
		}

		get hideSignature(): boolean {
			if (this.theme === ThemeEnum.COMPTE_ENTREPRISE) {
				return true;
			}

			return Boolean(this.$slots.default);
		}

		get secondaryLogo(): LogoInfo | undefined {
			return secondaryLogoMapping[this.theme];
		}

		get hasSecondaryLogo(): boolean {
			return Boolean(this.$slots.default || this.secondaryLogo);
		}

		get logoContainerComponent(): string {
			return this.homeLink ? 'RouterLink' : 'div';
		}

		get showDivider(): boolean {
			return Boolean(this.hasSecondaryLogo || this.service.title);
		}

		get showServiceSubTitle(): boolean {
			return Boolean(this.service.title && this.service.subTitle && !this.mobileVersion);
		}

		get dividerColor(): string {
			switch (this.theme) {
				case ThemeEnum.CNAM:
				case ThemeEnum.AMELI_PRO: {
					return tokens.colors.secondary;
				}

				case ThemeEnum.COMPTE_ENTREPRISE: {
					return '#cd545b'; // Brique 100 color
				}

				default: {
					return tokens.colors.primary;
				}
			}
		}

		get dividerDimensions(): Dimensions {
			const { xSmall, small, normal } = dividerDimensionsMapping;

			if (this.mobileVersion && this.hasSecondaryLogo) {
				return xSmall;
			}

			if (this.mobileVersion) {
				return small;
			}

			return normal;
		}

		get logoSize(): LogoSizeEnum {
			if (this.mobileVersion && this.hasSecondaryLogo) {
				return LogoSizeEnum.X_SMALL;
			}

			if (this.mobileVersion) {
				return LogoSizeEnum.SMALL;
			}

			return LogoSizeEnum.NORMAL;
		}
	}
</script>

<style lang="scss" scoped>
	.vd-header-brand-section {
		overflow: hidden;

		.vd-header-title {
			line-height: 1.45 !important;
		}

		.vd-compte-entreprise-title {
			font-weight: 700 !important;

			span {
				color: #cd545b;
			}
		}

		img,
		::v-deep img {
			width: auto;
			height: 100%;
			flex: none;
		}

		svg,
		.vd-header-home-link {
			flex: none;
		}
	}
</style>
