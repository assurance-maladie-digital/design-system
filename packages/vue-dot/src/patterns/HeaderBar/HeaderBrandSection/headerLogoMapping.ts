import { IndexedObject } from 'src/types';
import { locales } from './locales';
import { LogoInfo } from './types';

export const headerLogoMapping: IndexedObject<LogoInfo> = {
	logoAM: {
		src: require('../../../assets/logos/desktop/assurance-maladie.svg'),
		alt: locales.logoAssuranceMaladie
	},
	logoRisquePro: {
		src: require('../../../assets/logos/desktop/risque-pro.svg'),
		alt: locales.logoRisquePro
	}
};

export const secondaryLogoMapping: IndexedObject<LogoInfo> = {
	ameli: {
		src: require('../../../assets/logos/desktop/ameli.svg'),
		alt: locales.logoAmeli
	},
	ameliPro: {
		src: require('../../../assets/logos/desktop/ameli-pro.svg'),
		alt: locales.logoAmeliPro
	}
};
