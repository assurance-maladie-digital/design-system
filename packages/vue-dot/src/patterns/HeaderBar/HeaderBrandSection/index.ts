import HeaderBrandSection from './HeaderBrandSection.vue';

export default HeaderBrandSection;
