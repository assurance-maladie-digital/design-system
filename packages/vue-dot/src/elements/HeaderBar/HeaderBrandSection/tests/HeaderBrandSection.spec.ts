import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';
import { html } from '@/tests/utils/html';

import HeaderBrandSection from '../';

let wrapper: Wrapper<Vue>;

// Tests
describe('HeaderBrandSection', () => {
	it('renders correctly', () => {
		// Mount component
		wrapper = mountComponent(HeaderBrandSection);

		expect(html(wrapper)).toMatchSnapshot();
	});
});