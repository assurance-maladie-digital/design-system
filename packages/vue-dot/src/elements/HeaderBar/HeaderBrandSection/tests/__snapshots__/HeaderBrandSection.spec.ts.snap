// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBrandSection renders correctly 1`] = `
<div class="d-flex"><img src="" alt="Caisse Nationale d'Assurance Maladie">
  <div class="d-flex"><img src="" alt="Caisse Nationale d'Assurance Maladie"> <img src="" alt="Caisse Nationale d'Assurance Maladie"></div>
  <!---->
</div>
`;
